### PR TITLE
[AIRFLOW-2651] Add file system hooks with a common interface

### PR DESCRIPTION
**Pull request by jrderuiter**
### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-XXX
    - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a JIRA issue.

https://issues.apache.org/jira/browse/AIRFLOW-2651

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR adds a set of file system hooks (FsHooks) that implement a common interface for manipulating files on different types of file systems. For more details, see the JIRA issue above. Currently the PR implements FsHooks for FTP, HDFS3, S3, SFTP and the local file system. 

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

- tests/hooks/fs_hooks/test_base.py
- tests/hooks/fs_hooks/test_ftp.py
- tests/hooks/fs_hooks/test_hdfs3.py
- tests/hooks/fs_hooks/test_local.py
- tests/hooks/fs_hooks/test_s3.py
- tests/hooks/fs_hooks/test_sftp.py

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

TODO. 

### Documentation
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

TODO.

### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: airflow/hooks/fs_hooks/base.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+class FsHook(BaseHook):
+    """Base FsHook defining the FsHook interface and providing some basic
+       functionality built on this interface.
+    """
+
+    _conn_classes = {
+        'ftp': _FS_BASE_MODULE + '.ftp.FtpHook',
+        'hdfs': _FS_BASE_MODULE + '.hdfs3.Hdfs3Hook',
+        'local': _FS_BASE_MODULE + '.local.LocalFsHook',
```
Comment (ashb) [inline airflow/hooks/fs_hooks/base.py:42]: Local FS is not good practice and shouldn't be included -- we don't want people to get in to the habit of storing files on disk IMO

File: airflow/hooks/fs_hooks/base.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+    # Path manipulation methods.
+
+    # Join/split should be overriden for file systems that don't use
+    # posix path conventions.
+
+    @staticmethod
+    def join(path, *paths):
+        """Join one or more path components."""
+        return posixpath.join(path, *paths)
```
Comment (ashb) [inline airflow/hooks/fs_hooks/base.py]: Do these methods really add any value over just calling `posixpath.join` directly? The smaller the interface the easier it is to reason about.

File: airflow/hooks/fs_hooks/_fnmatch.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    """
+    name = os.path.normcase(name)
+    pat = os.path.normcase(pat)
+    return fnmatchcase(name, pat, sep=sep)
+
+
+@functools.lru_cache(maxsize=256, typed=True)
+def _compile_pattern(pat, sep):
+    if isinstance(pat, bytes):
+        pat_str = str(pat, 'ISO-8859-1')
```
Comment (ashb) [inline airflow/hooks/fs_hooks/_fnmatch.py]: why latin-1?

File: airflow/hooks/fs_hooks/_fnmatch.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+import os
+import posixpath
+import re
+import functools
+
+
+__all__ = ["filter", "fnmatch", "fnmatchcase", "translate"]
+
+
+def fnmatch(name, pat, sep=os.sep):
```
Comment (ashb) [inline airflow/hooks/fs_hooks/_fnmatch.py]: How is this different to fnmatch from the stdlib?

File: airflow/hooks/fs_hooks/s3.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
```
Comment (ashb) [inline airflow/hooks/fs_hooks/s3.py:1]: This shouldn't be a new hook - it should be methods added to the existing S3Hook.

File: airflow/hooks/fs_hooks/s3.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+        for child in self.get_conn().ls(parent_dir, detail=True):
+            if child['Key'] == path and \
+                    child['StorageClass'] == 'DIRECTORY':
+                return True
+
+        return False
+
+    def makedir(self, dir_path, mode=0o755, exist_ok=True):
+        self.makedirs(dir_path, mode=mode, exist_ok=exist_ok)
```
Comment (ashb) [inline airflow/hooks/fs_hooks/s3.py:82]: S3 doesn't have the concept of directories -- you can just write to paths at will.

File: airflow/hooks/fs_hooks/sftp.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
+# -*- coding: utf-8 -*-
```
Comment (ashb) [inline airflow/hooks/fs_hooks/sftp.py:1]: Again, this shouldn't be a new hook, it should be part of the existing sftp hook.

**Review by ashb — CHANGES_REQUESTED** [review]
There is a lot of new code here that duplicates existing hooks.

That is going to be confusing for on-going maintenance.

I haven't given a fuller review than the few comments at the moment as it's hard to see what is new and what is duplicated.

File: airflow/hooks/fs_hooks/_fnmatch.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        res = translate(pat, sep=sep)
+    return re.compile(res).match
+
+
+def filter(names, pat, sep=os.sep):
+    """Return the subset of the list NAMES that match PAT."""
+    result = []
+    pat = os.path.normcase(pat)
+    match = _compile_pattern(pat, sep=sep)
+    if os.path is posixpath:
```
Comment (Fokko) [inline airflow/hooks/fs_hooks/_fnmatch.py]: Maybe we can rewrite this a bit:
```
if os.path is posixpath:
    names = [os.path.normcase(name) for name in names]
names = [match(name) for name in names]
```
I like short code ;)

File: airflow/hooks/fs_hooks/base.py (hunk 1 above)
Comment (Fokko) [reply airflow/hooks/fs_hooks/base.py:42]: I agree partially with you there @ashb, but a local path could also be a network mounted disk.

File: airflow/hooks/fs_hooks/base.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+            root = pattern[:ind + 1]
+        else:
+            root = '.'
+
+        # Get all file paths.
+        all_paths = []
+        for base_dir, _, file_names in self.walk(root):
+            all_paths.extend(self.join(base_dir, f) for f in file_names)
+
+        # Filter using modified version of fnmatch.
```
Comment (Fokko) [inline airflow/hooks/fs_hooks/base.py]: Maybe shorten this to:
```
return fnmatch.filter(all_paths, pattern, sep=self.sep)
```

File: airflow/hooks/fs_hooks/hdfs3.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+                # Build connection.
+                self._conn = hdfs3.HDFileSystem(
+                    host=config.host or MyNone,
+                    port=config.port or MyNone,
+                    pars=pars,
+                    **extra_kws)
+
+        return self._conn
+
+    def disconnect(self):
```
Comment (Fokko) [inline airflow/hooks/fs_hooks/hdfs3.py:61]: Being autistic, but sometimes I see:
```
def disconnect(self):
    if self._conn is not None:
        self._conn.disconnect()
    self._conn = None
```
and
```
def disconnect(self):
    if self._conn is not None:
        self._conn.disconnect()
        self._conn = None
```

File: airflow/hooks/fs_hooks/local.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+import os
+import shutil
+
+from .base import FsHook
+
+
+class LocalFsHook(FsHook):
+    """Hook for interacting with local files on the local file system."""
+
+    sep = os.sep
```
Comment (Fokko) [inline airflow/hooks/fs_hooks/local.py:31]: This one isn't used

File: airflow/hooks/fs_hooks/base.py [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+        matches = fnmatch.filter(all_paths, pattern, sep=self.sep)
+
+        return matches
+
+
+class NotSupportedError(NotImplementedError):
```
Comment (Fokko) [inline airflow/hooks/fs_hooks/base.py:260]: Where's this one used?

File: airflow/hooks/fs_hooks/s3.py (hunk 5 above)
Comment (Fokko) [reply airflow/hooks/fs_hooks/s3.py:1]: I agree, otherwise these will diverge and people will get confused

File: airflow/hooks/fs_hooks/sftp.py (hunk 7 above)
Comment (Fokko) [reply airflow/hooks/fs_hooks/sftp.py:1]: I agree. I believe @NielsZeilemaker enriched the existing SFTP hook recently.

File: tests/hooks/fs_hooks/test_sftp.py [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+        """Tests the `makedir` method with an existing dir
+           and exist_ok = True.
+        """
+
+        self._mock_fs.exists.return_value = True
+
+        with self._mocked_hook as hook:
+            hook.makedirs('path/to/dir', mode=0o755, exist_ok=True)
+
+    # def test_glob(self):
```
Comment (milton0825) [inline tests/hooks/fs_hooks/test_sftp.py:193]: can you remove this

**Review by Fokko — CHANGES_REQUESTED** [review]
Hi @jrderuiter. Thanks for taking the time to contribute this to Airflow. I really like the concepts and it adds some structure that Airflow is currently lacking. @ashb has some valid comments, and I also added a few.

Apart from that, I think it would be also really valuable to add a bit of documentation, so that other contributors can also add support, for example GCS. cc @fenglu-g @kaxil 

Cheers, Fokko

**Comment by Fokko** [comment]
Can you also rebase onto master to resolve conflicts? :)

File: airflow/hooks/fs_hooks/base.py (hunk 1 above)
Comment (jrderuiter) [reply airflow/hooks/fs_hooks/base.py:42]: Besides this, the local hook can be useful for reading files locally and uploading them to a remote location. In that sense it mirrors some of the upload_file methods that exist now, but generalises by allowing files to be read anywhere by using different source/destination hooks, including the local file system.

File: airflow/hooks/fs_hooks/_fnmatch.py (hunk 4 above)
Comment (jrderuiter) [reply airflow/hooks/fs_hooks/_fnmatch.py]: It allows for recursive globs, which the standard fnmatch does not allow. We can also not support recursive globbing, which allows for a simpler codebase by allowing us to rely on the standard fnmatch.

File: airflow/hooks/fs_hooks/_fnmatch.py (hunk 3 above)
Comment (jrderuiter) [reply airflow/hooks/fs_hooks/_fnmatch.py]: This is the same as the stdlib I guess? That is where most of the fnmatch code originated from.

File: airflow/hooks/fs_hooks/base.py (hunk 2 above)
Comment (jrderuiter) [reply airflow/hooks/fs_hooks/base.py]: I agree that a smaller interface is better, however I wanted to keep the flexibility for supporting file systems that do not necessarily use posix-path style joins. If this is only different on windows, we can consider just using posixpath. (Is Airflow supposed to support windows?)

**Comment by jrderuiter** [comment]
Thanks @ashb and @Fokko for the comments. I think the discussion about how to integrate this with existing hooks is an important issue, but it is not quite clear for me how this would best be solved. 

The new SFTP hook implements most functionality of the existing hook, so dropping that hook from contrib would be possible, at the cost of changing the interface. I think this would be acceptable if the changes are integrated into a new major version of Airflow, rather than a new minor version. We can integrate any missing functionality into the new hook.

The existing S3 hook is a bit more difficult, as the new hook represents S3 as a file system, whilst the existing hook stays closer to the notion of S3 as a key-based blob store. As such, methods such as `select_key` and `list_prefixes` have no clear meaning in my proposed S3FsHook, as these aren't file system concepts. We could integrate some of these methods into the S3FsHook, but it would dilute the interface considerably, I fear. 

I think the two options we have are to:

1. Keep two separate hooks, one for S3 as a file system and one for S3 as a blob store.
2. Keep both interfaces, but separate the blob store interface into a sub-property of the hook to keep the interfaces separate. This would still require separating the existing S3Hook from the AwsHook base class, to avoid multiple inheritance (as we would need to inherit from both FsHook and AwsHook otherwise). S3Hook could still use the AwsHook through composition, as the only thing the AwsHook is really needed for is fetching the client and/or credentials. The only other option would be to make FsHook a mixin, but I fear that that would lead to dilution of the interface. 

I would be keen to hear your ideas on this.

**Comment by jrderuiter** [comment]
I would like to stress that the biggest advantage of having a common interface for file systems is that it becomes easier to write operators/hooks that read/write to different (combinations of) file systems. For example, by combining these file system hooks we can essentially copy files between any of the file systems (by using specific source/destination hooks), rather than being limited to copying from local to/from a specific file system.

File: airflow/hooks/fs_hooks/base.py (hunk 2 above)
Comment (jrderuiter) [reply airflow/hooks/fs_hooks/base.py]: We could also keep the methods internal and default to posixpath, allowing subclasses to override if necessary. This would keep the interface cleaner, at the cost of losing some agnosticity for the caller concerning the used file system (they may need to account for the separators of the concerned fs).

File: airflow/hooks/fs_hooks/s3.py (hunk 6 above)
Comment (jrderuiter) [reply airflow/hooks/fs_hooks/s3.py:82]: That's why makedirs is essentially a no-op. It does check for existence of the 'directory', to keep the idea of a 'file system' in line with the other hooks.

File: airflow/hooks/fs_hooks/s3.py (hunk 5 above)
Comment (jrderuiter) [reply airflow/hooks/fs_hooks/s3.py:1]: I don't necessarily agree, as the current S3 hook represents S3 more as a key-based blob store than a file system. See my comment below for more details. A compromise may be to combine the two interfaces, allowing the hook to be used as a file system or the current interface. This would bloat the interface somewhat though.

File: airflow/hooks/fs_hooks/base.py (hunk 2 above)
Comment (ashb) [reply airflow/hooks/fs_hooks/base.py]: All versions of windows since 7? Vista? have supported `/` in paths - i.e. `c:/foo.txt` works so we don't need to worry about this and can remove it.

**Comment by ashb** [comment]
Definitely not having two versions of the hooks is crucial.

For the case of S3: yes, s3 is not a filesystem. But the "fs"-like interfaces should be implemented as method on the existing S3 hook using the existing methods where possible. i.e. `walk()` would use `list_prefixes()` in someway. Anything else is just going to make long-term maintenance too hard.

Multiple inheritance is entirely possible in python, and having:

```
class S3Hook(FsHook, AwsHook):
```

would be reasonable in this case.

**Comment by bolkedebruin** [comment]
@jrderuiter I like the possibilities that this will deliver, but I think some architectural updates are required. The `lineage` improvements basically allow for the same kind of functionality and these changes will need to tie in with it. Maybe have a discussion offline or on th emaling list (or an improvement proposal) can speed this up?
